R-package/CMakeLists.txt

```diff
@@ -6,36 +6,32 @@ file(GLOB_RECURSE R_SOURCES
   ${CMAKE_CURRENT_LIST_DIR}/src/*.c)
 # Use object library to expose symbols
 add_library(xgboost-r OBJECT ${R_SOURCES})
-
-set(R_DEFINITIONS
+target_compile_definitions(xgboost-r
+  PUBLIC
   -DXGBOOST_STRICT_R_MODE=1
   -DXGBOOST_CUSTOMIZE_GLOBAL_PRNG=1
   -DDMLC_LOG_BEFORE_THROW=0
   -DDMLC_DISABLE_STDIN=1
   -DDMLC_LOG_CUSTOMIZE=1
   -DRABIT_CUSTOMIZE_MSG_
   -DRABIT_STRICT_CXX98_)
```

Comment:
These flags are unique to R and they make changes to `objxgboost`. The resulting `libxgboost.so` is substantially different from what the Python and JVM bindings need.

Reply:
To share a common library, we'd need to carefully extract R-specific code into a separate CMake target, so that the R flags only modify the separate target and not the common library. This may be nontrivial effort.

Reply:
For cran yes, for conda maybe no.  Let me take a look, see how much effort does it take.

Reply:
Can you review this PR as it is? Let's follow up with another PR for the common library discussion.